Add dependency mutation (#368)
gradle/gradle-profiler: README.md

A scenario can define changes that should be applied to the source before each build. You can use this to benchmark or profile an incremental build. The following mutations are available:

- `apply-build-script-change-to`: Add a statement to a Groovy or Kotlin DSL build script, init script or settings script. Each iteration adds a new statement and removes the statement added by the previous iteration.

> - `apply-dependency-change-to`: Add dependencies as projects to a Groovy or a Kotlin DSL build script. Each iteration adds a new combinations of projects as dependencies and removes the projects added by the previous iteration.

Reviewer: I think the name should be reflective to the fact that we are modifying _project_ dependencies. Something like `--apply-project-dependency-change-to`?

Author: Agreed. I would be happy to get a suggestions also for:
```
            projects-set-size = 10
            applied-projects-size = 5
```
since it feels like they don't tell exactly what they do (I already renamed that 2 times but I am still not happy)

Author: Renamed, also renamed `applied-projects-size` to `applied-projects-set-size`

        tasks = ["assemble"]

        apply-build-script-change-to = "build.gradle.kts"
        apply-project-dependency-change-to {
            files = ["build.gradle"]
            # Default number of applied-projects-count is 3.
            # Gradle-profiler will generate minimal number of projects to have a unique combination of project dependencies every run.

>             # Note: Number of projects is calculated as binomial coffiecient: "X choose applied-projects-count = iterations * files",

Reviewer: Could you clarify this a bit?

Author: I tried, let me know what you think

>             # due to that, higher the applied-projects-count is, less projects will be generated.

Reviewer: ```suggestion
            # Because of this the higher the applied-projects-count is the fewer projects will be generated.
```

Author: I removed that "because of this..." line, it seems confusing. If someone will want use some different non-default value, he/she can calculate values with provided formula.

>             applied-projects-count = 3

Reviewer: This term is still not meaningful for me. What is "applied" is the change itself, not the projects. What we want here is to say how many subproject dependencies the profiler is supposed to add. How about:

`project-dependency-count = 3`

or:

`dependency-count = 3`

or even simply:

`count = 3`

?

I think in the context even `count` makes perfect sense.

Author: Changed to dependency-count. For me it seems `count` does not hold enough information. And `dependency-count` is nice in case we will have also `apply-external-dependency-change-to` or smth similar in future

        tasks = ["assemble"]

        apply-build-script-change-to = "build.gradle.kts"
        apply-project-dependency-change-to {
            files = ["build.gradle"]

>             // Example: This will apply a unique combination of 5 projects from a set of 10 projects.

Author: Changed to 2.

>             // Example: This will apply a unique combination of 5 projects from a set of 10 projects.

Reviewer: 👍

            // Number of combinations can be calculated by typing "10 choosen 5" in your favorite search engine.
            // When both values are equal there is just one combination possible, so all
            // runs will have the same combination of the projects applied as dependency.
            // Default values are 10 and 5. 
            projects-set-size = 10

>             applied-projects-set-size = 5

Author: Number of projects is autogenerated now. User can specify `applied-projects-count`, default is 3. Also renamed that set size to `applied-projects-count`.  Explanation is still complicated but nevertheless

        apply-build-script-change-to = "build.gradle.kts"
        apply-project-dependency-change-to {
            files = ["build.gradle"]
            # Default number of applied-projects-count is 3.

>             # Gradle-profiler will generate minimal number of projects to have a unique combination of project dependencies every run.

Author: I think this is not easy to understand...

Reviewer: How about:
```suggestion
            # Gradle Profiler will simulate changes to project dependencies by generate some additional projects and then add a combination of project dependencies to every non-generated subprojects before each iteration.
            # The profiler will generate the minimal number of subprojects to allow for a unique combination of dependencies to be used for each iteration.
```
